llvm/include/llvm/ADT/APFloat.h

```diff
@@ -769,6 +777,246 @@ class DoubleAPFloat final : public APFloatBase {
   friend hash_code hash_value(const DoubleAPFloat &Arg);
 };
 
+class DFPFloat final : public APFloatBase {
+public:
+  /// \name Constructors
+  /// @{
+
+  DFPFloat(const fltSemantics &) {} // Default construct to +0.0
+  DFPFloat(const fltSemantics &, integerPart) {
+    assert(false && "Not Implemented");
+  }
+  DFPFloat(const fltSemantics &, uninitializedTag) {
+    assert(false && "Not Implemented");
+  }
+  DFPFloat(const fltSemantics &, const APInt &) {
+    assert(false && "Not Implemented");
+  }
+  DFPFloat(const DFPFloat &) { assert(false && "Not Implemented"); }
+  DFPFloat(DFPFloat &&) { assert(false && "Not Implemented"); }
+  ~DFPFloat() {}
+
+  /// @}
+
+  /// Returns whether this instance allocated memory.
+  bool needsCleanup() const { return partCount() > 1; }
+
+  /// \name Arithmetic
+  /// @{
+
+  opStatus add(const DFPFloat &, roundingMode);
+  opStatus subtract(const DFPFloat &, roundingMode);
+  opStatus multiply(const DFPFloat &, roundingMode);
+  opStatus divide(const DFPFloat &, roundingMode);
+
+  opStatus remainder(const DFPFloat &);
+
+  opStatus mod(const DFPFloat &);
+  opStatus fusedMultiplyAdd(const DFPFloat &, const DFPFloat &, roundingMode);
+  opStatus roundToIntegral(roundingMode);
+
+  opStatus next(bool nextDown);
+
+  /// @}
+
+  /// \name Sign operations.
+  /// @{
+
+  void changeSign();
+
+  /// @}
+
+  /// \name Conversions
+  /// @{
+
+  opStatus convert(const fltSemantics &, roundingMode, bool *);
+  opStatus convertToInteger(MutableArrayRef<integerPart>, unsigned int, bool,
+                            roundingMode, bool *) const;
+  opStatus convertFromAPInt(const APInt &, bool, roundingMode);
+  opStatus convertFromSignExtendedInteger(const integerPart *, unsigned int,
+                                          bool, roundingMode);
+  opStatus convertFromZeroExtendedInteger(const integerPart *, unsigned int,
+                                          bool, roundingMode);
+  Expected<opStatus> convertFromString(StringRef, roundingMode);
+  APInt bitcastToAPInt() const;
+  double convertToDouble() const;
+  float convertToFloat() const;
+
+  /// @}
+
+  bool operator==(const DFPFloat &) const = delete;
+
+  cmpResult compare(const DFPFloat &) const;
```

Comment:
Just a note that might be worth adding a comment for. Non-canonical values are unordered with respect to other (non-canonical or canonical) equal values per IEEE 754:2008 5.10, "Details of totalOrder predicate". It looks like `cmpResult` corresponds to 5.11, "Details of comparison predicates" which doesn't distinguish between differently encoded value representations.